<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>hubward &mdash; hubward 0.2.2 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="hubward 0.2.2 documentation" href="#" />
    <link rel="next" title="Installation" href="installation.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="hubward">
<h1><code class="docutils literal"><span class="pre">hubward</span></code><a class="headerlink" href="#hubward" title="Permalink to this headline">¶</a></h1>
<a class="reference external image-reference" href="https://travis-ci.org/daler/hubward"><img src="https://travis-ci.org/daler/hubward.svg" /></a>
<a class="reference external image-reference" href="http://bioconda.github.io"><img alt="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat-square" src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat-square" /></a>
<p>A common question when reading an interesting paper is, &#8220;how does relate to my
favorite gene locus?&#8221;. To address this question, usually we need to download
the supplemental data, figure out what format it&#8217;s in, convert it to some other
useful format, and visualize it alongside our own data.</p>
<p><cite>hubward</cite> helps manage this process, lowering the effort required to track,
manage, organize, and upload many studies for cross-comparison in the UCSC
Genome Browser.</p>
<p>Data are organized into <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/hgTrackHubHelp.html">track hubs</a> on the UCSC
Genome Browser, and the name <cite>hubward</cite> indicates the direction in which data
are moved into these track hubs. It can also refer to a direction in <a class="reference external" href="http://wiki.lspace.org/mediawiki/Hubwards">other
complex systems</a>.</p>
<p>The separate repository, <a class="reference external" href="https://github.com/daler/hubward-studies">hubward-studies</a>, contains examples of prepared
track hubs. Subsets of these can be combined into user-defined hubs, or can
serve as examples for preparing other studies.</p>
<p>Documentation can be found at <a class="reference external" href="https://daler.github.io/hubward">https://daler.github.io/hubward</a>.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://github.com/daler/hubward/raw/master/docs/images/hubward-poster.pdf">This poster</a> gives an overview of <cite>hubward</cite>, and provides a worked example.</p>
<p>Further details can be found in the <a class="reference external" href="https://daler.github.io/hubward">full documentation</a>.</p>
<p><cite>hubward</cite> uses the following concepts:</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">track:</th><td class="field-body">Data that can be represented as a single track in the UCSC Genome Browser.
Examples include a file of called peaks; read pileup from a single RNA-seq
sample; CNV scores for one sample; or anything that can be converted into
bigBed, bigWig, BAM, or VCF format.</td>
</tr>
<tr class="field-even field"><th class="field-name">study:</th><td class="field-body">A collection of tracks, typically all from the same published article.</td>
</tr>
<tr class="field-odd field"><th class="field-name">group:</th><td class="field-body">A collection of studies, typically related in some way.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="studies">
<h2>Studies<a class="headerlink" href="#studies" title="Permalink to this headline">¶</a></h2>
<p>The minimal definition of a <cite>hubward</cite> &#8220;study&#8221; is a directory with
a <cite>metadata.yaml</cite> file. In practice, the directory contains raw data and
conversion scripts. A study generally corresponds to data from a single
published paper, but this is not required.  The <cite>metadata.yaml</cite> file describes
and configures one or many <cite>tracks</cite> grouped together. These are uploaded to
a track hub as a single <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/hubQuickStartGroups.html#composite">composite track</a>.</p>
<p>The <cite>metadata.yaml</cite> file consists of several sections. The <cite>study</cite> section
stores bibliographic information. It is converted to HTML documentation and
added to the study&#8217;s configuration page in the UCSC Genome Browser.</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="l-Scalar-Plain">study</span><span class="p-Indicator">:</span>
  <span class="l-Scalar-Plain">reference</span><span class="p-Indicator">:</span> <span class="s">&#39;Ho,</span><span class="nv"> </span><span class="s">J.</span><span class="nv"> </span><span class="s">W.</span><span class="nv"> </span><span class="s">K.</span><span class="nv"> </span><span class="s">et</span><span class="nv"> </span><span class="s">al.</span><span class="nv"> </span><span class="s">Nature</span><span class="nv"> </span><span class="s">512,</span><span class="nv"> </span><span class="s">449-452</span><span class="nv"> </span><span class="s">(2014).&#39;</span>
  <span class="l-Scalar-Plain">PMID</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">25164756</span>
  <span class="l-Scalar-Plain">description</span><span class="p-Indicator">:</span> <span class="s">&#39;ENCODE</span><span class="nv"> </span><span class="s">predicted</span><span class="nv"> </span><span class="s">enhancers&#39;</span>
  <span class="l-Scalar-Plain">label</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">encode-enhancers</span>
  <span class="l-Scalar-Plain">processing</span><span class="p-Indicator">:</span> <span class="s">&quot;Downloaded</span><span class="nv"> </span><span class="s">data</span><span class="nv"> </span><span class="s">were</span><span class="nv"> </span><span class="s">converted</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">bigBed</span><span class="nv"> </span><span class="s">format&quot;</span>
</pre></div>
</div>
<p>The <cite>tracks</cite> section is a list, with one item for each track to be included in
the hub. Here is one such item in the <cite>tracks</cite> list:</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="l-Scalar-Plain">tracks</span><span class="p-Indicator">:</span>
  <span class="p-Indicator">-</span>
    <span class="l-Scalar-Plain">label</span><span class="p-Indicator">:</span> <span class="s">&quot;enhancers</span><span class="nv"> </span><span class="s">[K562]&quot;</span>
    <span class="l-Scalar-Plain">description</span><span class="p-Indicator">:</span> <span class="s">&quot;K562</span><span class="nv"> </span><span class="s">enhancers&quot;</span>
    <span class="l-Scalar-Plain">genome</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">hg19</span>
    <span class="l-Scalar-Plain">original</span><span class="p-Indicator">:</span> <span class="s">&quot;raw-data/p300_enhancers_K562.txt&quot;</span>
    <span class="l-Scalar-Plain">processed</span><span class="p-Indicator">:</span> <span class="s">&quot;processed-data/p300_enhancers_K562.bigbed&quot;</span>
    <span class="l-Scalar-Plain">script</span><span class="p-Indicator">:</span> <span class="s">&quot;scr/process.py&quot;</span>
    <span class="l-Scalar-Plain">source</span><span class="p-Indicator">:</span>
      <span class="l-Scalar-Plain">fn</span><span class="p-Indicator">:</span> <span class="s">&quot;comparative_enhancer_calls.tar.gz&quot;</span>
      <span class="l-Scalar-Plain">url</span><span class="p-Indicator">:</span> <span class="s">&quot;http://compbio.med.harvard.edu/modencode/webpage/enh_calls_final/comparative_enhancer_calls.tar.gz&quot;</span>
    <span class="l-Scalar-Plain">trackinfo</span><span class="p-Indicator">:</span>
       <span class="l-Scalar-Plain">tracktype</span><span class="p-Indicator">:</span> <span class="s">&quot;bigBed</span><span class="nv"> </span><span class="s">3&quot;</span>
       <span class="l-Scalar-Plain">visibility</span><span class="p-Indicator">:</span> <span class="s">&quot;dense&quot;</span>
       <span class="l-Scalar-Plain">itemRgb</span><span class="p-Indicator">:</span> <span class="s">&quot;on&quot;</span>
       <span class="l-Scalar-Plain">color</span><span class="p-Indicator">:</span> <span class="s">&quot;#FF0000&quot;</span>
    <span class="l-Scalar-Plain">type</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">bigbed</span>
</pre></div>
</div>
<p>The config file format and fields are described in detail later in the
documentation. In summary, this block defines the source data, an output file
to create, and a conversion script to create a bigBed file with features
colored red, for enhancers in K562 cells from the ENCODE project.  A logical
extension of this would be to include additional tracks for other cell lines in
this data set.</p>
<p>To process the data for a study, use:</p>
<div class="highlight-python"><div class="highlight"><pre>hubward process &lt;directory&gt;
</pre></div>
</div>
<p>where <cite>directory</cite> contains the <cite>metadata.yaml</cite> file. For each defined track, this will:</p>
<ul class="simple">
<li>ensure original data exist. If not, the <cite>source</cite> url is downloaded to the
<cite>source</cite> fn and extracted</li>
<li>ensure processed file exists and is up-to-date. If it is older than
<cite>original</cite> or older than <cite>script</cite>, the script is re-run.</li>
</ul>
</div>
<div class="section" id="groups">
<h2>Groups<a class="headerlink" href="#groups" title="Permalink to this headline">¶</a></h2>
<p>Multiple studies can be grouped together using a higher-level config file,
here called <cite>group.yaml</cite>. Each study can have multiple tracks; each group can
have multiple studies.</p>
<p>For example, if the path to the above <cite>metadata.yaml</cite> file is
<cite>encode/hg19/encode-enhancers</cite>, then that directory can be included in the
<cite>studies</cite> list so that the K562 enhancers track will be uploaded:</p>
<div class="highlight-yaml"><div class="highlight"><pre><span class="l-Scalar-Plain">group</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">encode</span>
<span class="l-Scalar-Plain">genome</span><span class="p-Indicator">:</span> <span class="l-Scalar-Plain">hg19</span>
<span class="l-Scalar-Plain">name</span><span class="p-Indicator">:</span> <span class="s">&quot;encodetracks&quot;</span>
<span class="l-Scalar-Plain">short_label</span><span class="p-Indicator">:</span> <span class="s">&quot;Supp.</span><span class="nv"> </span><span class="s">ENCODE&quot;</span>
<span class="l-Scalar-Plain">long_label</span><span class="p-Indicator">:</span> <span class="s">&quot;Supplemental</span><span class="nv"> </span><span class="s">ENCODE</span><span class="nv"> </span><span class="s">tracks&quot;</span>
<span class="l-Scalar-Plain">hub_url</span><span class="p-Indicator">:</span> <span class="s">&quot;http://localhost/encode/hg19/compiled.hub.txt&quot;</span>

<span class="l-Scalar-Plain">server</span><span class="p-Indicator">:</span>
  <span class="l-Scalar-Plain">hub_remote</span><span class="p-Indicator">:</span> <span class="s">&quot;/root/encode/hg19/compiled.hub.txt&quot;</span>
  <span class="l-Scalar-Plain">host</span><span class="p-Indicator">:</span> <span class="s">&quot;localhost&quot;</span>
  <span class="l-Scalar-Plain">user</span><span class="p-Indicator">:</span> <span class="s">&quot;www&quot;</span>
  <span class="l-Scalar-Plain">email</span><span class="p-Indicator">:</span> <span class="s">&quot;www@localhost&quot;</span>

<span class="l-Scalar-Plain">studies</span><span class="p-Indicator">:</span>
  <span class="p-Indicator">-</span> <span class="l-Scalar-Plain">encode/hg19/encode-enhancers</span>
</pre></div>
</div>
<p>To process all studies in a group, run:</p>
<div class="highlight-python"><div class="highlight"><pre>hubward process &lt;group YAML file&gt;
</pre></div>
</div>
<p>This processes all configured studies to ensure their output is up-to-date.</p>
<p>To create the track hub files and upload to a remote server, run:</p>
<div class="highlight-python"><div class="highlight"><pre>hubward upload &lt;group YAML file&gt;
</pre></div>
</div>
<p>After it runs, it will show the URL that can be used to load the hub into the
Genome Browser.</p>
</div>
<div class="section" id="workflow">
<h2>Workflow<a class="headerlink" href="#workflow" title="Permalink to this headline">¶</a></h2>
<p>To visualize a new dataset, the workflow is the following:</p>
<ol class="arabic simple">
<li>Write a <cite>metadata.yaml</cite> file and the corresponding scripts to perform
conversion.</li>
<li>Write a <cite>group config</cite> file file that groups together individual studies.</li>
<li>Run <cite>hubward process &lt;group config&gt;</cite>. This parses the group config file, and
for each defined study, parses its <cite>metadata.yaml</cite> file, downloads data if
needed, runs conversion scripts if necessary.</li>
<li>Run <cite>hubward upload &lt;group config&gt;</cite>. This builds the track hub config
files using the <cite>trackhub</cite> Python package, and uploads to the server
configured in the group config.</li>
</ol>
</div>
<div class="section" id="going-further">
<h2>Going further<a class="headerlink" href="#going-further" title="Permalink to this headline">¶</a></h2>
<p>Use <cite>hubward skeleton</cite> to create a template study including directories,
and a <cite>metadata-builder.py</cite> script to aid in programmatic generation of
<cite>metadata.yaml</cite>.</p>
<p><cite>hubward</cite> includes many helper functions which can be imported into the
processing script.</p>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#conda"><cite>conda</cite></a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#docker">Docker</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="hubward.html">hubward package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="hubward.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="hubward.html#module-hubward.generate_config_from_schema">hubward.generate_config_from_schema module</a></li>
<li class="toctree-l2"><a class="reference internal" href="hubward.html#module-hubward.liftover">hubward.liftover module</a></li>
<li class="toctree-l2"><a class="reference internal" href="hubward.html#module-hubward.log">hubward.log module</a></li>
<li class="toctree-l2"><a class="reference internal" href="hubward.html#module-hubward.models">hubward.models module</a></li>
<li class="toctree-l2"><a class="reference internal" href="hubward.html#module-hubward.utils">hubward.utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="hubward.html#module-hubward.version">hubward.version module</a></li>
<li class="toctree-l2"><a class="reference internal" href="hubward.html#module-hubward">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#types-of-contributions">Types of Contributions</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#get-started">Get Started!</a></li>
<li class="toctree-l2"><a class="reference internal" href="contributing.html#pull-request-guidelines">Pull Request Guidelines</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="authors.html#development-lead">Development Lead</a></li>
<li class="toctree-l2"><a class="reference internal" href="authors.html#contributors">Contributors</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="history.html#id1">0.2.2 (2016-01-20)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id2">0.2.1 (2015-12-05)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id3">0.2.0 (2015-10-24)</a></li>
<li class="toctree-l2"><a class="reference internal" href="history.html#id4">0.1.0 (2014-01-11)</a></li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span>Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span>Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span>Search Page</span></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><code class="docutils literal"><span class="pre">hubward</span></code></a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#studies">Studies</a></li>
<li><a class="reference internal" href="#groups">Groups</a></li>
<li><a class="reference internal" href="#workflow">Workflow</a></li>
<li><a class="reference internal" href="#going-further">Going further</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
      <li>Next: <a href="installation.html" title="next chapter">Installation</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2014, Ryan Dale.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.7</a>
      
      |
      <a href="_sources/index.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>